### PR TITLE
Adding optional field to NetworkWatcherFlowLog

### PR DESCRIPTION
**Pull request by mansi-verma-1510**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [ ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [ ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [ ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories")
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogList.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
               "targetResourceId": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Network/networkSecurityGroups/vm5-nsg",
               "targetResourceGuid": "00000000-0000-0000-0000-000000000000",
               "storageId": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Storage/storageAccounts/wzstorage002",
+              "enabledFilteringCriteria": "srcIP=158.255.7.8 || dstPort=56891",
```
Comment (ramoka178) [inline specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogList.json:21]: >enabledFilteringCriteria [](http://example.com/codeflow?start=15&length=24)

Same comment from a different PR on the same issue.
[https://github.com/Azure/azure-rest-api-specs-pr/pull/18515](https://github.com/Azure/azure-rest-api-specs-pr/pull/18515)

This seems to be very tricky to get the correct input for the user.

enabledFilteringCriteria is said to be for filtering flowlogs. Can it go on the query parameters as $filter ?

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogCreate.json [hunk 2]
```diff
@@ -10,6 +10,7 @@
       "properties": {
         "targetResourceId": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/desmondcentral-nsg",
         "storageId": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.Storage/storageAccounts/nwtest1mgvbfmqsigdxe",
+        "enabledFilteringCriteria": "srcIP=158.255.7.8 || dstPort=56891",
```
Comment (ramoka178) [inline specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogCreate.json:13]: Why is this needed in a PUT request? PUT should just create a flowLog resource. 

enabledFilteringCriteria is said to be for filtering flowlogs

Same in response too.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogGet.json [hunk 3]
```diff
@@ -17,6 +17,7 @@
           "targetResourceId": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Network/networkSecurityGroups/vm5-nsg",
           "targetResourceGuid": "00000000-0000-0000-0000-000000000000",
           "storageId": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Storage/storageAccounts/wzstorage002",
+          "enabledFilteringCriteria": "srcIP=158.255.7.8 || dstPort=56891",
```
Comment (ramoka178) [inline specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogGet.json:20]: >"enabledFilteringCriteria [](http://example.com/codeflow?start=10&length=25)

Why is this needed on a point GET request on a single resource?

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogUpdateTags.json [hunk 4]
```diff
@@ -23,6 +23,7 @@
           "targetResourceId": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/desmondcentral-nsg",
           "targetResourceGuid": "00000000-0000-0000-0000-000000000000",
           "storageId": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.Storage/storageAccounts/nwtest1mgvbfmqsigdxe",
+          "enabledFilteringCriteria": "srcIP=158.255.7.8 || dstPort=56891",
```
Comment (ramoka178) [inline specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogUpdateTags.json:26]: >enabledFilteringCriteria [](http://example.com/codeflow?start=11&length=24)

Why is this needed here ?

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogList.json (hunk 1 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogList.json:21]: The NRP includes checks to verify the format of the passed value in this field. We are expecting an expression representing some conditions using relational and logical operators. Case sensitivity and extraneous spaces are disregarded in these specifications. Additional details and formatting specifications will be outlined in the documentation.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogCreate.json (hunk 2 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogCreate.json:13]: We expect this condition to be passed during flowlogs creation step. we will collect logs according to this property set on the flowlogs resource.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogGet.json (hunk 3 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogGet.json:20]: customer may want to know what filtering conditions have been set for a flowlogs resource. So, we return this information too in the GET call for flowlogs resource.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogUpdateTags.json (hunk 4 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/examples/NetworkWatcherFlowLogUpdateTags.json:26]: We want to return all the properties related to flowlogs resource in the update response too..

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json [hunk 5]
```diff
@@ -3192,6 +3192,10 @@
           "description": "ID of the storage account which is used to store the flow log.",
           "type": "string"
         },
+        "enabledFilteringCriteria": {
+          "description": "Optional condition to filter flowlogs",
```
Comment (TimLovellSmith) [inline specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3196]: You should have a fuller description that documents the syntax of this string, or links to related docs, otherwise people won't know how to use it.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json [hunk 6]
```diff
@@ -3192,6 +3192,10 @@
           "description": "ID of the storage account which is used to store the flow log.",
           "type": "string"
         },
+        "enabledFilteringCriteria": {
```
Comment (TimLovellSmith) [inline specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3195]: Why are you calling it 'enabledFilteringCriteria'?

One way I'm puzzled by it - If you had any filtering criteria, why wouldn't you want to enable them?

Comment (TimLovellSmith) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3195]: Also, are there any special reasons you went with the mini language that you, or other processors of, will have to parse / emit? 

While it might seem pretty cool or simple that you can just write a nice powerful expression manually, (like the power of regex) it may not be as simple or pleasant to use in other contexts, like authoring ARM templates (where they might have to use a lot of 'concat'), or parsing your GET responses to fully analyze the current filtering state of the resource.

Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3195]: We are a logging service. By default, we capture all the logs of customer's network traffic. This is the scenario which you can call as filtering 'disabled'. In this case, customer will not pass anything to this field value.
--
The reason we are using this expression is due to the nature of logs. We capture 5 tuple information (srcIP, destIP, srcPort, DestPort and Protocol). So, filtering criteria could be a compound condition of these field. Mathematical expression consisting of logical and relational operators best suits this. This is also the way filtering is provided in our competitive services in google cloud and aws. This has been discussed internally within team with Eng Leads and PMs.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json (hunk 5 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3196]: we will publish the documentation about this when the feature is out to the customers for use.

Comment (TimLovellSmith) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3196]: Hmm, how will you get your spec and SDKs, all updated to point to it? Would an aka.ms placeholder link be a good aid in getting that done smoothly?

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json (hunk 6 above)
Comment (TimLovellSmith) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3195]: So its being called 'enabled' because its absence implies logging is 'disabled'? Or because the things matching the filer are 'enabled' for logging (logged), instead of 'disabled'? How can we make this clearer? Updating the description to explain whether you're filtering in, or filtering out, and what happens by default seems like the least you could do. 

I still wonder if its worth trying to come up with a better name, but I'm not sure I have any yet.

**Comment by TimLovellSmith** [comment]
@Joanna-Yang-Art was it the wrong target branch?

**Comment by GuptaVertika** [comment]
@TimLovellSmith , This branch doesnot contain 2024-03-01 version: using this branch: https://github.com/Azure/azure-rest-api-specs/tree/guptavertika/release-network-Microsoft.Network-2024-03-01

**Comment by TimLovellSmith** [comment]
Is there a replacement PR I should prioritize reviewing?

**Comment by GuptaVertika** [comment]
@mansi-verma-1510 Please open a new PR on branch https://github.com/Azure/azure-rest-api-specs/tree/guptavertika/release-network-Microsoft.Network-2024-03-01

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json (hunk 6 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3195]: Sure i'll update the description of the field and make it more verbose.
Regarding field name, we have already discussed this with the PMs.

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json (hunk 5 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3196]: I am not sure I understand this comment.
Are you asking us to add link to the documentation in description of the field?

File: specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json (hunk 6 above)
Comment (mansi-verma-1510) [reply specification/network/resource-manager/Microsoft.Network/stable/2024-01-01/networkWatcher.json:3195]: Updated
